File: 238-product-of-array-except-self/238-product-of-array-except-self.c


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* productExceptSelf(int* nums, int numsSize, int* returnSize){
    
    int* res = (int*)calloc(numsSize, sizeof(int));
    *returnSize = numsSize;
    
    int prefix = 1;
    //Prefix
    for(int i = 0; i < numsSize; i++){  
        res[i] = prefix;
        prefix *= nums[i];
    }
    
    int postfix = 1;
    //Postfix and results
    for(int i = numsSize-1; i >= 0; i--){
        res[i] *= postfix;
        postfix *= nums[i];
    }

    
    return res;
}

File: 238-product-of-array-except-self/238-product-of-array-except-self.go
func productExceptSelf(nums []int) []int {
    
    var prefix int = 1
    var postfix int = 1
    n := len(nums)
    
    res := make([]int, n)
    
    for i := 0; i < n; i++ {
        res[i] = 1
    } 
    
    for idx, _ := range nums {
        res[idx] = prefix
        prefix *= nums[idx]
    }
    
    for i := n - 1; i >= 0; i-- {
        res[i] *= postfix
        postfix *= nums[i]
    }
    
    return res
    
}